improver_tests/synthetic_data/test_generate_metadata.py

```diff
@@ -0,0 +1,556 @@
+    default_cube = generate_metadata()
+    cube.coord("time").bounds = None
+    cube.coord("forecast_period").bounds = None
+
+    assert cube == default_cube
+
+
+def test_set_frt():
+    """ Tests cube generated with specified forecast reference time and the rest of the
+    values set as default values """
+    frt = datetime(2017, 1, 1, 0, 0)
+    cube = generate_metadata(frt=frt)
+
+    assert iris_time_to_datetime(cube.coord("forecast_reference_time"))[0] == frt
+    assert cube.coord("forecast_period").points > 0
+
+    # Assert that no other values have unexpectedly changed by returning changed values
+    # to defaults and comparing against default cube
+    default_cube = generate_metadata()
+
+    cube.coord("forecast_reference_time").points = default_cube.coord(
+        "forecast_reference_time"
+    ).points
+    cube.coord("forecast_period").points = default_cube.coord("forecast_period").points
+
+    assert cube == default_cube
+
+
+def test_set_ensemble_members():
+    """ Tests cube generated with specified number of ensemble members """
+    ensemble_members = 4
+    cube = generate_metadata(ensemble_members=ensemble_members)
+
+    assert cube.ndim == 3
+    assert cube.shape == (ensemble_members, NPOINTS_DEFAULT, NPOINTS_DEFAULT)
+    assert cube.coords()[0].name() == "realization"
+
+    # Assert that cube shape is different from default cube shape but metadata unchanged
+    _check_cube_shape_different(cube)
+
+
+@pytest.mark.parametrize("ensemble_members", (0, 1))
+def test_disable_ensemble(ensemble_members):
+    """ Tests cube generated without realizations dimension """
+    cube = generate_metadata(ensemble_members=ensemble_members)
+
+    assert cube.ndim == 2
+    assert cube.shape == (NPOINTS_DEFAULT, NPOINTS_DEFAULT)
+
+    spatial_grid_values = SPATIAL_GRID_ATTRIBUTE_DEFAULTS[SPATIAL_GRID_DEFAULT]
+    assert cube.coords()[0].name() == spatial_grid_values["y"]
+    assert cube.coords()[1].name() == spatial_grid_values["x"]
+
+    # Assert that cube shape is different from default cube shape but metadata unchanged
+    _check_cube_shape_different(cube)
+
+
+@pytest.mark.parametrize(
+    "cube_type", ("variable", "percentile", "probability", "other")
+)
+@pytest.mark.parametrize(
+    "spp__relative_to_threshold", ("above", "below", "between", None),
+)
+def test_leading_dimension(cube_type, spp__relative_to_threshold):
+    """ Tests cube generated with leading dimension specified using percentile and
+    probability flags, and different values for spp__relative_to_threshold """
+    if cube_type == "probability" and spp__relative_to_threshold == "between":
+        # Tests that error is raised when both percentile and probability set to True
+        msg = "spp__relative_to_threshold"
+
+        with pytest.raises(ValueError, match=msg):
+            generate_metadata(
+                cube_type=cube_type,
+                spp__relative_to_threshold=spp__relative_to_threshold,
```

Comment:
A couple of things here. One, this isn't testing new functionality, but rather functionality within set_up_test_cubes which is already tested [here](https://github.com/metoppv/improver/blob/043d2ec33db5f6704e5f4a8d18dbad953cbafdc1/improver_tests/synthetic_data/test_set_up_test_cubes.py#L679). So I don't think we need the "between" test.

Secondly, I don't actually think this behaviour is entirely consistent with our code as we have the ability to generate "between" threshold probabilties, which you can find [here](https://github.com/metoppv/improver/blob/043d2ec33db5f6704e5f4a8d18dbad953cbafdc1/improver/between_thresholds.py#L196). That's a bit of an oversight / bug in the set_up_test_cubes code, though we are not actually using the between threshold diagnostics anywhere at the moment, so it is not a problem at present. Probably something to note down and return to if and when we ever need to.

Reply:
Removed the "between" test [here](https://github.com/metoppv/improver/pull/1326/commits/a8ae93b36fa2d7574acb2ab8f36b20f272d2a85f#diff-dfb229535f6ef5488fe1cf6ea49f02aeL334).

That's interesting, thank you for pointing it out.